lib/bundler/gem_helper.rb

```diff
@@ -98,7 +98,7 @@ def rubygem_push(path)
         allowed_push_host = @gemspec.metadata["allowed_push_host"]
         gem_command += " --host #{allowed_push_host}" if allowed_push_host
       end
-      unless allowed_push_host || Pathname.new("~/.gem/credentials").expand_path.file?
+      unless allowed_push_host || Bundler.user_home.join(".gem/credentials").file?
```

Comment:
Okay, as I suspected. That means we don't want to use `Bundler.user_home` here. In fact, this makes me think that all this logic should be inside the CompactIndex cache writer, and not inside `Bundler.user_home`, because we actively don't want to use a temp directory for either writing a settings file or looking for `.gem/credentials`. We specifically and only want to use a temp directory for the compact index files.